File: packages/schemas/Result.py
"""
Result + prediction sub-models — Pydantic source of truth.

prediction.json in the export pack is the canonical serialization of
PredictionPayload. It must be timestamp-free so its hash is stable across
replays (ARCHITECTURE.md §6 rule 4).

See README.md for drift policy.
"""
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class OffTargetHit(BaseModel):
    model_config = ConfigDict(frozen=True)

    sequence: str       # 20nt off-target sequence
    position: int       # 0-based position in backbone
    mismatches: int     # number of mismatches vs guide (≤4 for CFD scoring)
    cfd_score: float    # CFD score per Doench 2016 mismatch matrix, [0, 1]


class GuideCandidate(BaseModel):
    model_config = ConfigDict(frozen=True)

    sequence: str                           # 20nt protospacer
    pam: str                                # 3nt PAM (e.g. "AGG")
    position: int                           # 0-based start in backbone (+ strand)
    strand: Literal["+", "-"]
    on_target_score: float                  # Doench RS2, [0, 1]
    off_target_count: int                   # total ≤4-mismatch hits
    off_target_top_hits: list[OffTargetHit] # top-5 by CFD score
    bystander_warnings: list[str]           # adjacent C/A in editing window


class PredictionPayload(BaseModel):
    """
    Timestamp-free payload — its canonical JSON hash is replay-stable.
    Do NOT add datetime, uuid, or run_id fields here (ARCHITECTURE.md §6 rule 4).
    """

    model_config = ConfigDict(frozen=True)

    guides: list[GuideCandidate]
    summary: dict   # type: ignore[type-arg]  # top score, mean off-target, etc.


class Result(BaseModel):
    model_config = ConfigDict(frozen=True)

    run_id: UUID
    prediction: PredictionPayload
    # export_pack_ref is a StorageRef shape: {"bucket": str, "path": str}.
    # Typed as dict to avoid cross-schema import; validated at the app layer.
    export_pack_ref: dict[str, str] | None = None  # StorageRef shape
    export_pack_sha256: str | None = None           # SHA-256 hex of zip bytes

File: packages/schemas/Result.ts
/**
 * Result + prediction sub-models — TypeScript mirror of Result.py.
 * Pydantic is the source of truth. See README.md for drift policy.
 *
 * prediction.json must be timestamp-free (ARCHITECTURE.md §6 rule 4).
 * Do NOT add Date, run_id, or any non-deterministic fields to PredictionPayload.
 */

import type { Sha256, StorageRef } from "./ResearchObject";

export type OffTargetHit = {
  readonly sequence: string; // 20nt
  readonly position: number; // 0-based in backbone
  readonly mismatches: number; // ≤4
  readonly cfd_score: number; // [0, 1]
};

export type GuideCandidate = {
  readonly sequence: string; // 20nt protospacer
  readonly pam: string; // 3nt PAM
  readonly position: number; // 0-based start (+ strand)
  readonly strand: "+" | "-";
  readonly on_target_score: number; // Doench RS2, [0, 1]
  readonly off_target_count: number;
  readonly off_target_top_hits: readonly OffTargetHit[]; // top-5 by CFD score
  readonly bystander_warnings: readonly string[];
};

/**
 * Timestamp-free payload. Its canonical JSON hash is replay-stable.
 * Never add Date, UUID, or run_id fields here.
 */
export type PredictionPayload = {
  readonly guides: readonly GuideCandidate[];
  readonly summary: Record<string, unknown>; // top score, mean off-target, etc.
};

export type Result = {
  readonly run_id: string; // UUID
  readonly prediction: PredictionPayload;
  readonly export_pack_ref: StorageRef | null; // populated when L5 finishes
  readonly export_pack_sha256: Sha256 | null; // SHA-256 of zip bytes
};
